src/schema/document/core.rs

```diff
@@ -0,0 +1,310 @@
+use std::collections::{BTreeMap, HashMap, HashSet};
+use std::net::Ipv6Addr;
+
+use common::DateTime;
+use serde_json::Map;
+
+use crate::schema::document::{
+    DeserializeError, DocumentAccess, DocumentDeserialize, DocumentDeserializer,
+};
+use crate::schema::field_type::ValueParsingError;
+use crate::schema::field_value::FieldValueIter;
+use crate::schema::{Facet, Field, FieldValue, NamedFieldDocument, Schema, Value};
+use crate::tokenizer::PreTokenizedString;
+
+/// Tantivy's Document is the object that can
+/// be indexed and then searched for.
+///
+/// Documents are fundamentally a collection of unordered couples `(field, value)`.
+/// In this list, one field may appear more than once.
+#[derive(Clone, Debug, serde::Serialize, serde::Deserialize, Default)]
+pub struct Document {
+    field_values: Vec<FieldValue>,
+}
+
+impl DocumentAccess for Document {
+    type Value<'a> = &'a Value;
+    type FieldsValuesIter<'a> = FieldValueIter<'a>;
+
+    fn iter_fields_and_values(&self) -> Self::FieldsValuesIter<'_> {
+        FieldValueIter(self.field_values.iter())
+    }
+}
+
+impl DocumentDeserialize for Document {
+    fn deserialize<'de, D>(mut deserializer: D) -> Result<Self, DeserializeError>
+    where D: DocumentDeserializer<'de> {
+        let mut field_values = Vec::with_capacity(deserializer.size_hint());
+
+        while let Some((field, value)) = deserializer.next_field()? {
+            field_values.push(FieldValue::new(field, value));
+        }
+
+        Ok(Self { field_values })
+    }
+}
+
+impl From<Vec<FieldValue>> for Document {
+    fn from(field_values: Vec<FieldValue>) -> Self {
+        Self { field_values }
+    }
+}
+
+impl PartialEq for Document {
+    fn eq(&self, other: &Self) -> bool {
+        // super slow, but only here for tests
+        let convert_to_comparable_map = |field_values: &[FieldValue]| {
```

Comment:
If we iterate like in `segment_writer` it would be faster

Reply:
Probably, this was from the original code so I didn't bother changing it.